File: src/main/java/leetcode/lists/hot100/LeetCode394.java
package leetcode.lists.hot100;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/12 00:10
 */
public class LeetCode394 {

    public String decodeString(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        return dfs(s, 0)[0];
    }

    public String[] dfs(String s, int i) {
        int sum = 0;
        StringBuilder sb = new StringBuilder();
        while (i < s.length()) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                sum = sum * 10 + Integer.parseInt(String.valueOf(s.charAt(i)));
            } else if (c == '[') {
                // 遇到[, 即可拆分成子问题， 定义好返回值，基于子问题的返回值继续求解
                String[] temp = dfs(s, i + 1);
                i = Integer.parseInt(temp[0]);
                while (sum > 0) {
                    sb.append(temp[1]);
                    sum--;
                }
            } else if (c == ']') {
                return new String[]{String.valueOf(i), sb.toString()};
            } else {
                sb.append(c);
            }
            i++;
        }
        return new String[]{sb.toString()};
    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode394.go
package hot100

func decodeString(s string) string {
	if len(s) == 0 {
		return ""
	}
	res, _ := decodeStringHelper(s, 0)
	return res
}

func decodeStringHelper(s string, i int) (string, int) {
	sum := 0
	var str string
	for i < len(s) {
		if IsNumber(s[i]) {
			//go study
			sum = sum*10 + int(s[i]-'0')
		} else if s[i] == '[' {
			substr, index := decodeStringHelper(s, i+1)
			i = index
			for sum > 0 {
				str += substr
				sum--
			}
		} else if s[i] == ']' {
			return str, i
		} else {
			str += string(s[i])
		}
		i++
	}
	return str, i
}
